docs/modules/ROOT/pages/explanations/decisions/nextcloud.adoc

```diff
@@ -0,0 +1,64 @@
+= VSHN Managed Nextcloud on Kubernetes
+
+== Problem
+
+We need to provide Nextcloud on Kubernetes with the following features:
+
+* Possibility to Single Sign-On with Keycloak
+* Nextcloud Office with Collabora on backend
+* PostgreSQL as database
+* Extensions management
+* full lifecycle of an application: deploy, update, backup, restore, monitoring, scaling, etc.
+
+
+== Solutions
+
+- helm chart available for Nextcloud: https://github.com/nextcloud/helm
+- Writing our own provider-nextcloud 
+
+=== Analysis - Helm Chart
+
+The Nextcloud Helm Chart is a good starting point for our offering. It is actively maintained and has a lot of features. It is also the official Helm Chart for Nextcloud.
+
+Nextcloud provides also: https://github.com/nextcloud/all-in-one docker image, which contains a lot of features including Office package.
+
+For backup and restore there is necessity to use a separate solution like k8up.
+For extension management we need to provide a custom solution.
+There is another helm chart for Office extension: https://open.greenhost.net/stackspin/nextcloud (not official)
+
+Advantages::
+
+* Official Helm Chart
+* Actively maintained
+* Supports various databases, metrics, persistance, ingress, some basic rbac
+
+Disadvantages::
+
+* We still need to cover a lot of topics manually:
+** Maintenance
+** Backup and restore
+** Extension management
+** Office package deployment and integration
+
+=== Analysis - Custom Provider
+
+Writing our own provider would give us the most flexibility. We could implement all the features we need and integrate them into our existing solutions. We could probably handle more features than the Helm Chart.
```

Comment:
Creating a custom provider to deploy a few Kubernetes objects seems pretty overkill. Provider-helm can do that much better. We should rather look into forking or creating our helm chart first, if that is necessary.

We've already tried the custom deploy provider approach in the beginning of schedar and it was not feasible. We should build on top and combine things that already exist.

Crossplane providers are usually intended to be very broad in functionality and then through compositions they get the actual business logic applied. So if we would have the need for a provider in the future it should not deploy nextcloud, but rather configure it via its API. It should manage things like:
* user
* extensions
* configurations -> e.g. oauth configs, so that we could automagically attach a vshnkeycloak if it exists
* etc.

There's an important distinction between the responsibilities of a Crossplane provider and a composition. Most providers connect to existing services and provide a means of configuring them. Typically, for each API call there's a single managed resource exposed by the provider. They normally don't provision a service from scratch. The composition then brings the actual business logic and combines the managed resource to so that the service gets created.

If at any point we need to talk to nextcloud's API (for whatever reason) I would strongly suggest not to that in the comp functions but rather move that to the provider-nextcloud. For the reasons stated above.